### PR TITLE
Add Notifications

### PR DESCRIPTION
**Pull request by dx24013**
Add notifications when the menu shows and hides. Specifically: SideMenuWillShow, SideMenuDidShow, SideMenuWillHide, and SideMenuDidHide.

### CODE REVIEW

**Comment by jonkykong** [comment]
Hi @dx24013,

Thank you for your suggestion.

Have you checked out the demo project? It shows you how you can catch these events by inspecting the `viewWillAppear`, etc. events in the `UISideMenuNavigationController` subclass.

Personally, I have avoided using Notifications because there can be delays in their posting. For responsive UI I find this to be unacceptable.

**Comment by jonkykong** [comment]
Another complication with this solution is that it may cause unexpected reactions from other view controllers elsewhere in the hierarchy.

An alternative solution that solve both this complication and ensure real-time responsiveness would be delegate methods. A bonus could be that no delegate needs to be assigned; it simply checks for protocol conformance and calls the methods if it does.

**Comment by dx24013** [comment]
Thanks, I will close this pull request and see about using delegate.

**Comment by NikKovIos** [comment]
I think the notifications would be a good feature. Even it can have a delays, the programmer should have such a tool when using your library. In notification description you can notice, that it possible a delays and manage the responsibility about using notifications to the developer.

In my project the notifications would be the best way to achieve my task. Better than delegate.

**Comment by jonkykong** [comment]
@NikKovIos this PR was suggested before the delegate methods were added. The delegate methods are the right way to go about this. They provide the same functionality as notifications but without any delay. Can you provide reasons for why notifications are better than a delegate?

If there's one thing I've learned about building a free library it's that many developers don't take the responsibility you suggest and instead look to the library owner to fix/solve their problems. Anything I can do to stop that from happening is preferred.

**Comment by NikKovIos** [comment]
@jonkykong of course it is your choice about what to implement and what not. There are advantages and disadvantages for every feature. Imho the way to build a free library is to give as much api as i can. Managing issues don't spend much time (in case of theme if this topic), because good documentation sieve the biggest part of potential issues.

**Comment by jonkykong** [comment]
@NikKovIos I disagree building as much as possible. More code = more problems, more confusion over what does what. The vast majority of developers seem pleased with the basics. Managing issues takes a few minutes out of every day, and that adds up and is a distraction.

Outside of this, I'm still interested to hear your logical reasons of why notifications are better than delegates.

**Comment by NikKovIos** [comment]
The biggest benefit of notifications over delegates is that to implement notifications you don't need to think about controllers/view structure. Just implement and the works. And in my project i use storyboard with Side menu navigation controller with PageViewController in. And i need to manage the opening a menu from one of ViewController of the PageViewController.

**Comment by jonkykong** [comment]
The delegate design does not require you to specify a view controller as the delegate, it will always attempt to call the methods if they exist.

With the delegate approach, you need to be aware of the controllers/view/stack structure.

With the notification approach, you need to be aware of the controllers/view/stack structure to ensure that you don't perform certain action when the controller is out of view. You also need to ensure you call appropriate actions back on the main thread.

Notifications appear to have more cons.

**Comment by NikKovIos** [comment]
First: `The delegate design does not require you to specify a view controller as the delegate, it will always attempt to call the methods if they exist.` - wait a minute. Maybe i don't know, but the delegate must be defined explicitly like `UISideMenuNavigationController().sideMenuDelegate = someViewController`. If you don't define it and just write and extension for class with `UISideMenuNavigationControllerDelegate`, a delegate methods wouldn't be called.

Second: `With the notification approach, you need to be aware of the controllers/view/stack structure to ensure that you don't perform certain actions when the controller is out of view`. This is not the truth because you can remove the observer in viewWillDisappear if you need but commonly it's not needed. The thing, that you wrote about main thread is on developer responsibility and not in touch of this topic.

In conclusion i could resume that that the notifications is a separate tool. It has adv and disadv but this is a power thing and can't be replaced with the delegate only.

**Comment by jonkykong** [comment]
1. If you check the code, you _can_ specify the delegate, but I took the extra step of automatically assuming a default delegate of the presenting view controller for additional ease. It could technically backfire, but I think it made it one step easier.

2. Unfortunately, even though it is the developer's responsibility to call back on the main thread, I will still get questions from new developers who need help. Regardless, this is still additional code that needs to be written to handle this use case specific to notifications that does not apply to delegates. So, with notifications you must add code that ensures you're on the main thread, as well adds/removes observers for the controller when it is displayed/hidden. This is clearly more work than setting a delegate, which could simply be updated as each new view controller comes into view. However as I earlier explained, SideMenu smartly makes an assumption so you don't even have to do that.

I believe this all still clearly demonstrates that the delegate approach is the better solution, and notifications would be redundant and inferior in performance.
